A87-mtls-spiffe-support.md

```diff
@@ -0,0 +1,266 @@
+}
+class SpiffeId {
+   String getTrustDomain();
+   String getPath();
+}
+class SpiffeBundle {
+   Map<String, List<X509Certificates>> getBundleMap();
+}
+```
+
+For the xDS functionality described above, the `XdsX509TrustManager`
+will gain the ability to be instantiated with a SPIFFE trust bundle
+map.  In this case, it will use the SPIFFE trust bundle certificates as
+trusted roots.  If the `XdsX509TrustManager` is instantiated using CA
+certificates (existing functionality), then it'll continue to use them
+exactly as it does today.  The new APIs will look like this:
+
+```java
+class XdsTrustManagerFactory {
+   XdsTrustManagerFactory(Map<String, List<X509Certificates>>);
+   XdsTrustManagerFactory(X509Certificates[]);
+}
+class XdsX509TrustManager {
+   XdsX509TrustManager(Map<String, XdsX509ExtendedTrustManager>);
+   XdsTrustManagerFactory(XdsX509ExtendedTrustManager);
+}
+```
+
+We'll also adjust existing hierarchies of `Watcher` and
+`CertificateProvider` to support the new SPIFFE trust bundle map.
+
+For the AdvancedTls case in Java, the API changes will look as follows.
+In AdvancedTlsX509TrustManager.java, the API must allow configuring and
+reloading a SPIFFE Bundle Map.  There are currently three functions that
+configure the root trust certificate
+
+```
+public void updateTrustCredentials(X509Certificate[] trustCerts) throws IOException,
+   GeneralSecurityException
+
+public void updateTrustCredentials(File trustCertFile) throws IOException,
+   GeneralSecurityException
+
+public Closeable updateTrustCredentials(File trustCertFile, long period, TimeUnit unit,
+   ScheduledExecutorService executor) throws IOException, GeneralSecurityException
+```
+
+We will add two similar new functions for updating the SPIIFE root of trust:
+
+```
+public void updateSpiffeTrustBundle(File spiffeBundleFile) throws IOException,
+   GeneralSecurityException
+
+public Closeable updateSpiffeTrustBundle(File spiffeBundleFile, long period, TimeUnit unit,
+   ScheduledExecutorService executor) throws IOException, GeneralSecurityException
+```
+
+The `checkTrusted` implementations will be updated to properly use the SPIFFE Bundles as described in this gRFC.
+
+#### Go
+
+In Go, we can leverage [the go spiffe
+package](https://pkg.go.dev/github.com/spiffe/go-spiffe/v2) for parsing and
+object abstractions.  Currently, similarly to C++ above, Go already supports
+[file watcher credential
+providers](https://github.com/grpc/grpc-go/blob/e0d191d8adcdd73aad084154769404dd2f6b0fc6/credentials/tls/certprovider/pemfile/watcher.go#L92C4-L99)
+with xDS. We can further modify these providers, specifically the file watcher,
+to allow them to be configured with and return SPIFFE bundles instead of only
+certificates.
+Specifically, the Provider is designed to return a
+[KeyMaterial](https://github.com/grpc/grpc-go/blob/e0d191d8adcdd73aad084154769404dd2f6b0fc6/credentials/tls/certprovider/provider.go#L91-L97)
+struct where we could easily add a trust bundle.
```

Comment:
Can you be more specific about the actual change proposed to the cert provider?  Are we simply adding a field to `KeyMaterial`?  What is its name/type?

Reply:
Done, PTAL